### PR TITLE
Split nullable context for types and for warnings.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@cston, @jcouv, @333fred Please review

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 1]
```diff
@@ -70,14 +72,15 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, ReportDiag
...
+                    nullableOption,
                     generalDiagnosticOption,
                     specificDiagnosticOptions,
                     out hasPragmaSuppression);
             }
             else
             {
                 reportAction = GetDiagnosticReport(d.Severity, d.IsEnabledByDefault, d.Id, d.WarningLevel, d.Location as Location,
-                    d.Category, warningLevelOption, generalDiagnosticOption, specificDiagnosticOptions, out hasPragmaSuppression);
+                    d.Category, warningLevelOption, nullableOption , generalDiagnosticOption, specificDiagnosticOptions, out hasPragmaSuppression);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs]: >  [](start = 66, length = 1)

Extra space.

File: src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs [hunk 2]
```diff
@@ -15,6 +15,33 @@ internal static partial class ErrorFacts
         private const string s_descriptionSuffix = "_Description";
         private static readonly Lazy<ImmutableDictionary<ErrorCode, string>> s_helpLinksMap = new Lazy<ImmutableDictionary<ErrorCode, string>>(CreateHelpLinks);
         private static readonly Lazy<ImmutableDictionary<ErrorCode, string>> s_categoriesMap = new Lazy<ImmutableDictionary<ErrorCode, string>>(CreateCategoriesMap);
+        public static readonly ImmutableHashSet<string> NullableFlowAnalysisWarnings;
+
+        static ErrorFacts()
+        {
+            ImmutableHashSet<string>.Builder builder = ImmutableHashSet.CreateBuilder<string>();
+            builder.Add(MessageProvider.Instance.GetIdForErrorCode((int)ErrorCode.WRN_ConvertingNullableToNonNullable));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Errors/ErrorFacts.cs]: >MessageProvider.Instance.GetIdForErrorCode((int)ErrorCode.WRN_ConvertingNullableToNonNullable) [](start = 24, length = 94)

Perhaps extra a local function `string getId(ErrorCode code)`.

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 3]
```diff
@@ -160,6 +207,18 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, ReportDiag
             }
 
             return report;
+
+            bool promoteToAnError()
+            {
+                Debug.Assert(report == ReportDiagnostic.Default);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs]: >Debug [](start = 16, length = 5)

Should this method also assert `generalDiagnosticOption == ReportDiagnostic.Error`?

File: src/Compilers/CSharp/Portable/Syntax/CSharpPragmaWarningStateMap.cs [hunk 4]
```diff
@@ -17,85 +41,135 @@ internal class CSharpPragmaWarningStateMap : AbstractWarningStateMap
...
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(currentNullableDirective.SettingToken.Kind());
+                    }
 
-                        if (!string.IsNullOrWhiteSpace(errorId))
-                        {
-                            // Update the state of this error code with the current directive state
-                            accumulatedSpecificWarningState = accumulatedSpecificWarningState.SetItem(errorId, directiveState);
-                        }
+                    foreach (string id in ErrorFacts.NullableFlowAnalysisWarnings)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/CSharpPragmaWarningStateMap.cs]: >foreach [](start = 20, length = 7)

Perhaps use `ImmutableDictionary<K, V>.SetItems()` to reduce the number of allocations.

File: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs [hunk 5]
```diff
@@ -24,9 +25,10 @@ internal static class CSharpDiagnosticFilter
         /// <param name="d">The input diagnostic</param>
         /// <param name="warningLevelOption">The maximum warning level to allow. Diagnostics with a higher warning level will be filtered out.</param>
         /// <param name="generalDiagnosticOption">How warning diagnostics should be reported</param>
+        /// <param name="nullableOption">Nullable option</param>
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs]: Could you add more detail than just the parameter name?

**Comment by 333fred** [comment]
Done review pass (commit 1)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-infrastructure Is there something wrong with the "WIP" check? It looks like it is blocking an ability to merge this PR.

**Comment by jasonmalinowski** [comment]
@AlekseyTs Not sure what happened, but I don't see any obvious reason it should have been blocked. People often get tripped up with "WIP" appearing in a comment text somewhere, but it's not that...

**Pull request merged**